### ISSUE TITLE
[Question] Does Nx support multiple package.json files?

### ISSUE THREAD

**Issue by dereklin** [nrwl/nx#113]
It seems like there is only one package.json in the root directory and all apps and libs are using the same dependencies.

I would like to have different dependencies per app or lib.  Can this be done with nx?  If so, how can I do that?

**Comment by sean-perkins** [comment]
You can create a different `package.json` file in each designated app. I would recommend referencing shared dependencies across all your apps, by pointing the package to the root package.json dependency.

For example in `apps/example/package.json` my set-up may look like this:

```
{
  "license": "SEE LICENSE IN <your-license-filename>",
  "repository": "<fill-your-repository-here>",
  "scripts": {},
  "dependencies": {
    "@angular/animations": "file:../../node_modules/@angular/animations",
    "@angular/common": "file:../../node_modules/@angular/common",
    "@angular/compiler": "file:../../node_modules/@angular/compiler",
    "@angular/core": "file:../../node_modules/@angular/core",
    "@angular/forms": "file:../../node_modules/@angular/forms",
    "@angular/http": "file:../../node_modules/@angular/http",
    "@angular/platform-browser": "file:../../node_modules/@angular/platform-browser",
    "@angular/platform-browser-dynamic": "file:../../node_modules/@angular/platform-browser-dynamic",
    "@angular/router": "file:../../node_modules/@angular/router",
    "@scoped/example-lib": "file:../../libs/example",
    "custom-lib-for-app": "^1.0.4"
  },
  "devDependencies": {
    "@angular/compiler-cli": "file:../../node_modules/@angular/compiler-cli",
    "typescript": "file:../../node_modules/typescript"
  }
}
```

**Comment by dereklin** [comment]
@sean-perkins Thanks for the info.  I am evaluating whether we should go with the monorepo route with nx.  Let's say after sometime my app1 is stable and becomes a legacy app, but my app2 is still active.

I think with what you have shown me, I can have app1 depend on Angular 5, while app2 can depend on Angular 7.  Right?

Now let's say I have a common component as lib1, used by app1 and app2.  I upgrade lib1 to use Angular 7, or I change lib1's behavior, what is the best way to keep app1 from affected by lib1's change?  Is there some kind of versioning system for lib1?  Or do I keep lib1 as legacy, and create a similar lib, lib11, for app2.  If so, my lib1 and lib11 will have duplicate code.

In short, in nx monorepo, how does one deal with mixing legacy apps with active apps?

**Comment by sean-perkins** [comment]
@dereklin 

Great questions all around. 

### Version control on external dependencies
You can absolutely have mixed versions of the external dependencies. You would just upgrade the path from referencing root's dependencies (the defaults for NX) to whatever fixed version (or future version) your app needs. 

### Version control on internal scoped dependencies
By default, you cannot manage mixed versions of the libs folders. These scoped packages are always the current instance of the libs modules. You could however do a few different techniques here.
- You could `npm pack` your desired lib. Then move the packed `.tgz` file into your app directory and reference the path in that app's package.json (_pitfalls_: Cannot manipulate source for this snapshot)
- You could publish your lib to a public/private npm package and reference that version in your app's package.json (_recommended_).
- You could experiment with having your app's package.json reference different git branches for different versions of your libs (would not recommend as a long term solution).

**Comment by dereklin** [comment]
@sean-perkins Thanks very much for the confirmation and suggested strategies.  I feel much more comfortable using nx.
